### PR TITLE
Class models: Enforce `extends Realm.Object`

### LINKED ISSUE #4412: Realm#constructor should throw if schema class doesn’t extend Realm.Object.

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

This closes #4412 by checking the prototype of constructors passed into the `Realm` constructor.

## ☑️ ToDos
* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [x] 📱 Check the React Native/other sample apps work if necessary
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

### CODE REVIEW

File: integration-tests/tests/src/tests/class-models.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      }).throws("Class 'Person' must extend Realm.Object");
+
+      // Mutate the name of the object schema to produce a more detailed error
+      Person.schema.name = "Foo";
+      expect(() => {
+        new Realm({ schema: [Person as any] });
+      }).throws("Class 'Person' (declaring 'Foo' schema) must extend Realm.Object");
+    });
+
+    it("is allowed", () => {
```
Comment (takameyer) [inline integration-tests/tests/src/tests/class-models.ts:64]: It's probably covered elsewhere in the tests, but the only case I could think of adding would be plain object schemas, just to ensure the changes made left that flow alone.

File: integration-tests/tests/src/tests/class-models.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  describe("as schema element", () => {
+    beforeEach(() => {
+      Realm.clearTestState();
+    });
+
+    it("fails without a schema static", () => {
+      class Person extends Realm.Object {}
+      expect(() => {
+        new Realm({ schema: [Person as any] });
+      }).throws("must have a 'schema' property");
```
Comment (kneth) [inline integration-tests/tests/src/tests/class-models.ts:33]: Is it possible to get the class name in the error message?

File: integration-tests/tests/src/tests/class-models.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    });
+
+    it("fails without a schema.properties static", () => {
+      class Person extends Realm.Object {
+        static schema = { name: "Person" };
+      }
+      expect(() => {
+        new Realm({ schema: [Person as any] });
+      }).throws("properties must be of type 'object'");
```
Comment (kneth) [inline integration-tests/tests/src/tests/class-models.ts:42]: Is it possible to get the class name in the error message?

**Review by kneth — APPROVED** [review]
Only minor comments - not blocking in any way

File: integration-tests/tests/src/tests/class-models.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    it("is allowed", () => {
+      class Person extends Realm.Object {
+        name!: string;
+        static schema: Realm.ObjectSchema = {
+          name: "Person",
+          properties: { name: "string" },
+        };
+      }
+      new Realm({ schema: [Person] });
```
[the remark this answers is not in the document]
Comment (kraenhansen) [reply integration-tests/tests/src/tests/class-models.ts:72]: This doesn't really add anything though? If `new Realm()` throws the tests fails, I see little need to expect / assert that?

[the remark this answers is not in the document]
Comment (takameyer) [reply integration-tests/tests/src/tests/class-models.ts:72]: If there's no assertion, do we need a test?

File: integration-tests/tests/src/tests/class-models.ts (hunk 1 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/class-models.ts:64]: Yeah. The tests for plain objects remain in the legacy test suite, I'm not moving these over in this PR, but they do pass.

File: integration-tests/tests/src/tests/class-models.ts (hunk 3 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/class-models.ts:42]: Sure, but that code is not related to the changes I'm making in this PR.
I thought about leaving out the test, but put it in for completeness.

File: integration-tests/tests/src/tests/class-models.ts (hunk 2 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/class-models.ts:33]: Sure, but that code is not really related to the changes I'm making in this PR.
I thought about leaving out the test, but put it in for completeness.

File: integration-tests/tests/src/tests/class-models.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (takameyer) [reply integration-tests/tests/src/tests/class-models.ts:72]: Ok...I did some reading and found this to be valid (challenging my assumptions :) ).  It's called as smoke test if it has no assertion.  If it's broken (smoking) then it failed the test 😆

**Review by takameyer — APPROVED** [review]

**Review by fronck — APPROVED** [review]

**Pull request merged**